src/Fields/broadcast.jl

```diff
@@ -132,7 +146,12 @@ end
     space2::AbstractSpace,
 )
     if space1 !== space2
-        error_mismatched_spaces(typeof(space1), typeof(space2))
+        if is_diagonalized_spaces(typeof(space1), typeof(space2)) &&
+           allow_mismatched_diagonalized_spaces()
+            warn_mismatched_spaces(typeof(space1), typeof(space2))
```

Comment:
should we still check if they are `==` equal?

Reply:
This is inside a `space1 !== space2` branch?

Reply:
`!==` is not `===`. We should check if they are `==`.

Reply:
Ah, right, agreed.

Reply:
Can you explain this further @simonbyrne or @charleskawczynski, given that the `if space1 !== space2` has not been changed by this PR? I thought we wanted this error to check that memory was not being used wastefully, so `===` for identical memory address makes sense.

Reply:
So there are 3 possible cases we should consider:
- spaces are `===`: we're good
- spaces are not `===`, but are `==` (i.e. they are both constructed from the same way). In this case, we should warn, but let it go
- spaces are not `==`: should be an error.

The extra question is how to define `==`: in this case, the easiest is probably to just check if the `local_geometry` fields are `==`?

Reply:
- [x]  spaces are `===`: we're good

- [ ] spaces are not `===`, but are `==` (i.e. they are both constructed from the same way). In this case, we should warn, but let it go

- [x]  spaces are not `==`: should be an error.

> The extra question is how to define `==`: in this case, the easiest is probably to just check if the `local_geometry` fields are `==`?

Right, currently our "`==`" is `is_diagonalized_spaces`:
```
is_diagonalized_spaces(::Type{S}, ::Type{S}) where {S <: AbstractSpace} = true

is_diagonalized_spaces(::Type, ::Type) = false
```

Is it insufficient to check at this type level? Or overkill?